core/src/main/java/org/elasticsearch/cluster/ClusterStateTaskExecutor.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.cluster;
+
+import java.util.Arrays;
+import java.util.List;
+
+public interface ClusterStateTaskExecutor<T> {
+    /**
+     * Update the cluster state based on the current state and the given tasks. Return the *same instance* if no state
+     * should be changed.
+     */
+    Result execute(ClusterState currentState, List<T> tasks) throws Exception;
+
+    /**
+     * indicates whether this task should only run if current node is master
+     */
+    default boolean runOnlyOnMaster() {
+        return true;
+    }
+
+    class Result {
```

Comment:
@bleskes I agree on documentation, but I'm also pondering making it more explicit by having it be a `Map<T, Throwable>`. What do you think? I'm thinking something like this?

```diff
diff --git a/core/src/main/java/org/elasticsearch/cluster/ClusterStateTaskExecutor.java b/core/src/main/java/org/elasticsearch/cluster/ClusterStateTaskExecutor.java
index 861b924..6d5d11c 100644
--- a/core/src/main/java/org/elasticsearch/cluster/ClusterStateTaskExecutor.java
+++ b/core/src/main/java/org/elasticsearch/cluster/ClusterStateTaskExecutor.java
@@ -18,15 +18,16 @@
  */
 package org.elasticsearch.cluster;
 
-import java.util.Arrays;
+import java.util.Collections;
 import java.util.List;
+import java.util.Map;
 
 public interface ClusterStateTaskExecutor<T> {
     /**
      * Update the cluster state based on the current state and the given tasks. Return the *same instance* if no state
      * should be changed.
      */
-    Result execute(ClusterState currentState, List<T> tasks) throws Exception;
+    Result<T> execute(ClusterState currentState, List<T> tasks) throws Exception;
 
     /**
      * indicates whether this task should only run if current node is master
@@ -35,19 +36,18 @@ public interface ClusterStateTaskExecutor<T> {
         return true;
     }
 
-    class Result {
+    class Result<T> {
         final public ClusterState resultingState;
-        final public List<Throwable> failures;
+        final public Map<T, Throwable> failures;
 
-        public Result(ClusterState resultingState, int numberOfTasks) {
+        public Result(ClusterState resultingState) {
             this.resultingState = resultingState;
-            failures = Arrays.asList(new Throwable[numberOfTasks]);
+            failures = Collections.EMPTY_MAP;
         }
 
-        public Result(ClusterState resultingState, List<Throwable> failures) {
+        public Result(ClusterState resultingState, Map<T, Throwable> failures) {
             this.resultingState = resultingState;
             this.failures = failures;
         }
     }
-
 }
diff --git a/core/src/main/java/org/elasticsearch/cluster/ClusterStateUpdateTask.java b/core/src/main/java/org/elasticsearch/cluster/ClusterStateUpdateTask.java
index 17c4635..9fe41b7 100644
--- a/core/src/main/java/org/elasticsearch/cluster/ClusterStateUpdateTask.java
+++ b/core/src/main/java/org/elasticsearch/cluster/ClusterStateUpdateTask.java
@@ -41,9 +41,9 @@ abstract public class ClusterStateUpdateTask implements ClusterStateTaskConfig,
     }
 
     @Override
-    final public Result execute(ClusterState currentState, List<Void> tasks) throws Exception {
+    final public Result<Void> execute(ClusterState currentState, List<Void> tasks) throws Exception {
         ClusterState result = execute(currentState);
-        return new Result(result, tasks.size());
+        return new Result<>(result);
     }
 
     /**
diff --git a/core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataMappingService.java b/core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataMappingService.java
index 44e3853..ec1e9a8 100644
--- a/core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataMappingService.java
+++ b/core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataMappingService.java
@@ -81,9 +81,9 @@ public class MetaDataMappingService extends AbstractComponent {
 
     class RefreshTaskExecutor implements ClusterStateTaskExecutor<RefreshTask> {
         @Override
-        public Result execute(ClusterState currentState, List<RefreshTask> tasks) throws Exception {
+        public Result<RefreshTask> execute(ClusterState currentState, List<RefreshTask> tasks) throws Exception {
             ClusterState newClusterState = executeRefresh(currentState, tasks);
-            return new Result(newClusterState, tasks.size());
+            return new Result<>(newClusterState);
         }
     }
 
@@ -221,9 +221,9 @@ public class MetaDataMappingService extends AbstractComponent {
 
     class PutMappingExecutor implements ClusterStateTaskExecutor<PutMappingClusterStateUpdateRequest> {
         @Override
-        public Result execute(ClusterState currentState, List<PutMappingClusterStateUpdateRequest> tasks) throws Exception {
+        public Result<PutMappingClusterStateUpdateRequest> execute(ClusterState currentState, List<PutMappingClusterStateUpdateRequest> tasks) throws Exception {
             List<String> indicesToClose = new ArrayList<>();
-            ArrayList<Throwable> failures = new ArrayList<>(tasks.size());
+            Map<PutMappingClusterStateUpdateRequest, Throwable> failures = new HashMap<>();
             try {
                 // precreate incoming indices;
                 for (PutMappingClusterStateUpdateRequest request : tasks) {
@@ -250,13 +250,12 @@ public class MetaDataMappingService extends AbstractComponent {
                 for (PutMappingClusterStateUpdateRequest request : tasks) {
                     try {
                         currentState = applyRequest(currentState, request);
-                        failures.add(null);
                     } catch (Throwable t) {
-                        failures.add(t);
+                        failures.put(request, t);
                     }
                 }
 
-                return new Result(currentState, failures);
+                return new Result<>(currentState, failures);
             } finally {
                 for (String index : indicesToClose) {
                     indicesService.removeIndex(index, "created for mapping processing");
diff --git a/core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java b/core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java
index c97c55d..e32ec30 100644
--- a/core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java
+++ b/core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java
@@ -395,7 +395,7 @@ public class InternalClusterService extends AbstractLifecycleComponent<ClusterSe
             toExecute.stream().forEach(task -> task.listener.onNoLongerMaster(task.source));
             return;
         }
-        ClusterStateTaskExecutor.Result result;
+        ClusterStateTaskExecutor.Result<T> result;
         long startTimeNS = System.nanoTime();
         try {
             List<T> inputs = toExecute.stream().map(tUpdateTask -> tUpdateTask.task).collect(Collectors.toList());
@@ -410,20 +410,20 @@ public class InternalClusterService extends AbstractLifecycleComponent<ClusterSe
                 logger.trace(sb.toString(), e);
             }
             warnAboutSlowTaskIfNeeded(executionTime, source);
-            result = new ClusterStateTaskExecutor.Result(previousClusterState, Collections.nCopies(toExecute.size(), e));
+            Map<T, Throwable> failures =
+                    toExecute.stream().collect(Collectors.toMap(updateTask -> updateTask.task, updateTask -> e));
+            result = new ClusterStateTaskExecutor.Result<>(previousClusterState, failures);
         }
-        assert result.failures.size() == toExecute.size();
 
         ClusterState newClusterState = result.resultingState;
         final ArrayList<UpdateTask<T>> proccessedListeners = new ArrayList<>();
         // fail all tasks that have failed and extract those that are waiting for results
-        for (int i = 0; i < toExecute.size(); i++) {
-            final UpdateTask<T> task = toExecute.get(i);
-            final Throwable failure = result.failures.get(i);
+        for (UpdateTask<T> updateTask : toExecute) {
+            final Throwable failure = result.failures.get(updateTask.task);
             if (failure == null) {
-                proccessedListeners.add(task);
+                proccessedListeners.add(updateTask);
             } else {
-                task.listener.onFailure(task.source, failure);
+                updateTask.listener.onFailure(updateTask.source, failure);
             }
         }
 
diff --git a/core/src/test/java/org/elasticsearch/cluster/ClusterServiceIT.java b/core/src/test/java/org/elasticsearch/cluster/ClusterServiceIT.java
index 60e7fb2..7de963f 100644
--- a/core/src/test/java/org/elasticsearch/cluster/ClusterServiceIT.java
+++ b/core/src/test/java/org/elasticsearch/cluster/ClusterServiceIT.java
@@ -755,10 +755,10 @@ public class ClusterServiceIT extends ESIntegTestCase {
             private AtomicInteger counter = new AtomicInteger();
 
             @Override
-            public Result execute(ClusterState currentState, List<Task> tasks) throws Exception {
+            public Result<Task> execute(ClusterState currentState, List<Task> tasks) throws Exception {
                 tasks.forEach(task -> task.execute());
                 counter.addAndGet(tasks.size());
-                return new Result(currentState, tasks.size());
+                return new Result<>(currentState);
             }
 
             @Override
diff --git a/test-framework/src/main/java/org/elasticsearch/test/cluster/TestClusterService.java b/test-framework/src/main/java/org/elasticsearch/test/cluster/TestClusterService.java
index 3845a71..5225228 100644
--- a/test-framework/src/main/java/org/elasticsearch/test/cluster/TestClusterService.java
+++ b/test-framework/src/main/java/org/elasticsearch/test/cluster/TestClusterService.java
@@ -191,12 +191,12 @@ public class TestClusterService implements ClusterService {
             logger.debug("failed [{}], no longer master", source);
             return;
         }
-        ClusterStateTaskExecutor.Result result;
+        ClusterStateTaskExecutor.Result<T> result;
         ClusterState previousClusterState = state;
         try {
             result = executor.execute(previousClusterState, Arrays.asList(task));
         } catch (Exception e) {
-            result = new ClusterStateTaskExecutor.Result(previousClusterState, Arrays.asList(e));
+            result = new ClusterStateTaskExecutor.Result<>(previousClusterState, Collections.singletonMap(task, e));
         }
         if (result.failures.get(0) != null) {
             listener.onFailure(source, new ElasticsearchException("failed to process cluster state update task [" + source + "]",
```